src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs

```diff
@@ -374,7 +374,10 @@ private RegexNode ScanRegex()
                         break;
 
                     case '$':
-                        AddUnitType(UseOptionM() ? RegexNode.Eol : RegexNode.EndZ);
+                        if (UseOptionA())
+                            AddUnitType(UseOptionM() ? RegexNode.AnyEol : RegexNode.AnyEndZ);
+                        else
+                            AddUnitType(UseOptionM() ? RegexNode.Eol : RegexNode.EndZ);
                         break;
 
                     case '.':
```

Comment:
@stephentoub regarding "should it not also change the newlines that . maps to?", does this look right, assuming that we are allowing '\n', '\r' and '\r\n' ?

Reply:
The Unicode spec I linked recommends that `.` can match `\r\n` as if it was a single character:

> Where the 'arbitrary character pattern' matches a newline sequence, it must match all of the newline sequences, and \u{D A} (CRLF) should match as if it were a single character. (The recommendation that CRLF match as a single character is, however, not required for conformance to RL1.6.)

To achieve that is more complicated than the `[^\r\n]`. It suggests this (including all the newline characters):

`(?:\u{D A}|(?!\u{D A})[\u{A}-\u{D}\u{85}\u{2028}\u{2029}]`

Reply:
The pattern above is missing a paren, and I don't understand the hyphen. I think in .NET that would be `(?:\r\n}|(?!\r\n)[\r\n\u0085\u2028\u2029])`

Reply:
> > Do other implementations follow this standard in general?
> 
> As noted in https://www.regular-expressions.info/dot.html (section "Line Break Characters") -- it varies. According to this, Java is missing \f (formfeed - \u0C) for example. It does seem that Unicode is a safe standard to choose from.

tl;dr: Yes, agree, TR18 seems perfect:

1. Nobody can say .NET made an arbitrary, quick decision and missed something obvious
2. Easier for new C# developers to on-board from other languages

Thanks for finding this standard.  I was unaware of it, despite reading Freidl's _Mastering Regular Expressions_.  I searched through my PDF copy of the book, and he doesn't mention TR18.  I guess his book was last updated in 2006, and TR18 was published in 2000, so maybe it's taken time for it to become popular.  Amusingly, [Oracle's java.util.regex documentation](https://docs.oracle.com/javase/8/docs/api/java/util/regex/package-summary.html) does not mention TR18 but instead says "Go read _Mastering Regular Expressions_ for help using this library" 🤣 

> #### Related Documentation
> An excellent tutorial and overview of regular expressions is Mastering Regular Expressions, Jeffrey E. F. Friedl, O'Reilly and Associates, 1997.

Jokes aside, Rust's [regex 1.1.0 crate](https://docs.rs/crate/regex/1.1.0/source/UNICODE.md), RogueWave's [Internationalization module](https://docs.roguewave.com/en/sourcepro/11/html/i18nref/rwuregularexpression.html), and the eponymous ICU Regular Expression Engine all support TR18 Level 1, at minimum.  The package in the R CRAN repository, [stringi](https://cran.r-project.org/web/packages/stringi/index.html), internally uses C bindings to the ICU C Regex Engine.  It looks like Python 3.7 considered replacing the `re` module with a `regex` module that supports ICU as well.  

In looking through the .NET implementation, it seems like the one feature necessary for full TR18 compliance would be "character class intersection".  That should be a separate issue.

Reply:
> In looking through the .NET implementation, it seems like the one feature necessary for full TR18 compliance would be "character class intersection". That should be a separate issue.

There's more. Eg., we are also missing symmetric difference (`[a-g~~b-h] `), and it looks like we are missing some properties as well, e.g I noticed `\p{LC}` is (accidentally?) missing. I only looked briefly though, I think it would be worth doing a more careful look and opening a separate issue(s) for discrepancies.

Reply:
Incidentally `(?:\r\n}|(?!\r\n)[\r-\n\u0085\u2028\u2029])` will presumably bypass Stephen's ASCII optimizations to character class matching, even though the pattern might be otherwise ASCII. That might justify some tuning later.

Reply:
> will presumably bypass Stephen's ASCII optimizations to character class matching

Most of it will still apply, e.g. we'll still generate a fast lookup table for ASCII, and only fall back to the slower path if the actual character being examined isn't ASCII.  And it'll only happen if you opt-in to this AnyNewLine feature.

Reply:
Yes, and also: The Level 1 of the specification seems most solid, whereas Level 2 and Level 3 are undergoing change.  Level 3 is proposed to be completely removed in the next version, subject to vendor feedback.  They're basically asking, "Is this too crazy for anyone to implement?"

They've seemed to have taken feedback from EcmaScript Tc39 as some basic guideline on what's reasonable, as part of JavaScript's major contribution to humanity in the last decade: making sure you can Regex match emojis via RegEx character class:

```javascript
const reRgiEmoji = /\p{RGI_Emoji}/u;
```